### PR TITLE
[v3.6.0] Fix shadow and light errors

### PR DESCRIPTION
**Pull request by troublemaker52025**
Re: #

### Changelog

* 

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by troublemaker52025** [comment]
@cocos-robot run test cases

**Review by jk20012001 — APPROVED** [review]

File: cocos/core/renderer/scene/spot-light.ts [hunk 1]
```diff
@@ -240,6 +241,17 @@ export class SpotLight extends Light {
...
 
+    /**
+     * @en get or set this light has Cast Shadow Objects
+     * @zh 获取或者设置灯光有没有可以投射阴影的物体
+     */
+    get hasCastShadowObjects () {
+        return this._hasCastShadowObjects;
+    }
+    set hasCastShadowObjects (val) {
+        this._hasCastShadowObjects = val;
```
Comment (pandamicro) [inline cocos/core/renderer/scene/spot-light.ts]: Why give public API to set this ? Does user need to use it

File: native/tools/tojs/scene.ini [hunk 2]
```diff
@@ -81,7 +81,8 @@ skip = Pass::[getBlocks],
        Node::[setLayerPtr setUIPropsTransformDirtyCallback rotate$ setUserData getUserData getChildren rotateForJS setScale$ setRotation$ setRotationFromEuler$ setPosition$ isActiveInHierarchy setActiveInHierarchy setActiveInHierarchyPtr setRTS$ findComponent findChildComponent findChildComponents addComponent removeComponent getComponent getComponents getComponentInChildren getComponentsInChildren checkMultipleComp getEventProcessor dispatchEvent hasEventListener getUIProps getPosition getRotation getScale getEulerAngles getForward getUp getRight getWorldPosition getWorldRotation getWorldScale getWorldMatrix getWorldRS getWorldRT],
        Camera::[screenPointToRay screenToWorld worldToScreen worldMatrixToScreen syncCameraEditor getMatView getMatProj getMatProjInv getMatViewProj getMatViewProjInv],
        NodeUiProperties::[getUITransformComp setUITransformComp getUIComp setUIComp], # not impl
-       ProgramLib::[destroyInstance]
+       ProgramLib::[destroyInstance],
+       SpotLight::[hasCastShadowObjects]
```
Comment (pandamicro) [inline native/tools/tojs/scene.ini]: Why skip an in-exist API ?

File: cocos/core/pipeline/render-additive-light-queue.ts [hunk 3]
```diff
@@ -54,6 +54,13 @@ interface IAdditiveLightPass {
 
 const _lightPassPool = new Pool<IAdditiveLightPass>(() => ({ subModel: null!, passIdx: -1, dynamicOffsets: [], lights: [] }), 16);
 
+interface InstancedOrBatchedLightPass {
```
Comment (pandamicro) [inline cocos/core/pipeline/render-additive-light-queue.ts]: Just reuse `IAdditiveLightPass` and `_lightPassPool`, only use `dynamicOffsets` and `lights` for instanced or batched queue

File: cocos/core/pipeline/render-additive-light-queue.ts [hunk 4]
```diff
@@ -54,6 +54,13 @@ interface IAdditiveLightPass {
...
+    dynamicOffsets: number[];
+    lights: number[];
+}
+
+const _instancedOrBatchedLightPassPool = new Pool<InstancedOrBatchedLightPass>(() => ({ dynamicOffsets: [], lights: [] }), 16);
```
Comment (pandamicro) [inline cocos/core/pipeline/render-additive-light-queue.ts]: Reuse `_lightPassPool `

File: cocos/core/renderer/scene/spot-light.ts (hunk 1 above)
Comment (pandamicro) [reply cocos/core/renderer/scene/spot-light.ts]: And why Directional Light doesn't need this

**Review by pandamicro — CHANGES_REQUESTED** [review]
Please also modify cpp files accordingly

File: cocos/core/renderer/scene/spot-light.ts (hunk 1 above)
Comment (troublemaker52025) [reply cocos/core/renderer/scene/spot-light.ts]: I thought about it carefully:
* If there is no castModel in the scene, after enabling the scene shadow and spotlight shadow options, shadowflow should initialize this framebuffer and clear it.  It is reasonable to generate a framebuffer at this time, the framebuffer is reasonable, and it is reasonable to enable receiveshadow by default in the model to sample in the shder
* mainlight is the execution of the above process.  The purpose of the flag here is to skip the framebuffer generation of the spot light when the scene does not have a castModel, and skipping the shadowmap sampling is trick and optimize.

File: cocos/core/pipeline/render-additive-light-queue.ts (hunk 3 above)
Comment (troublemaker52025) [reply cocos/core/pipeline/render-additive-light-queue.ts]: sure

File: cocos/core/renderer/scene/spot-light.ts (hunk 1 above)
Comment (troublemaker52025) [reply cocos/core/renderer/scene/spot-light.ts]: if mainlight need this optimiztion，also mainlight need add `hasCastObjects` flag

Comment (jk20012001) [reply cocos/core/renderer/scene/spot-light.ts]: we should make sure that if spot shadow is enabled and no cast objects, shadow texture still need to be created and initialize to 1.0, not use default-texture

**Review by pandamicro — APPROVED** [review]

**Review by jk20012001 — APPROVED** [review]

**Pull request merged**